### PR TITLE
feat: system-wide dfx

### PR DESCRIPTION
**Pull request by ericswanson-dfinity**
# Description

Example: Run `dfx start` outside of a dfx project
```
$ ls
$ dfx-wip start
There is no project-specific network 'local' because there is no project (no dfx.json).
Using the default definition for the 'local' shared network because /Users/ericswanson/.config/dfx/networks.json does not exist.
Local server configuration:
  bind address: 127.0.0.1:8000
  bitcoin: disabled
  canister http: disabled
  replica:
    subnet type: Application
  data directory: /Users/ericswanson/Library/Application Support/org.dfinity.dfx/network/local
  scope: shared
```

Similar example: shared dfx.json exists, but doesn't define the network
```
$ echo "{}" >/Users/ericswanson/.config/dfx/networks.json
$ dfx-wip start                                     
There is no project-specific network 'local' because there is no project (no dfx.json).
Using the default definition for the 'local' shared network because /Users/ericswanson/.config/dfx/networks.json does not define it.
Local server configuration:
  bind address: 127.0.0.1:8000
  bitcoin: disabled
  canister http: disabled
  replica:
    subnet type: Application
  data directory: /Users/ericswanson/Library/Application Support/org.dfinity.dfx/network/local
  scope: shared
```

Example: run `dfx start` in a dfx project, where dfx.json defines the local network
```
$ jq .networks.local dfx.json
{
  "bind": "localhost:8000",
  "type": "ephemeral"
}
$ dfx-wip start
Found project-specific network 'local' in /Users/ericswanson/w/dd/dfx.json
WARN: Project-specific networks are deprecated and will be removed after February 2023.
Local server configuration:
  bind address: 127.0.0.1:8000
  bitcoin: disabled
  canister http: disabled
  replica:
    subnet type: Application
  data directory: /Users/ericswanson/w/dd/.dfx/network/local
  scope: project
```

Fixes https://dfinity.atlassian.net/browse/SDK-421

# How Has This Been Tested?

Added e2e tests

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

### CODE REVIEW

File: src/dfx/src/lib/network/id.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+use anyhow::Context;
+use fn_error_context::context;
+use serde::{Deserialize, Serialize};
+use time::OffsetDateTime;
+
+#[derive(Deserialize, Serialize, Debug)]
+struct Id {
+    created: OffsetDateTime,
+}
```
Comment (sesi200) [inline src/dfx/src/lib/network/id.rs:12]:
```suggestion
#[derive(Deserialize, Serialize, Debug)]
struct NetworkMetadata {
    created: OffsetDateTime,
}
```
This has a confusing name. How can a struct that only contains a timestamp be an identifier?

**Review by sesi200 — COMMENTED** [review]
LGTM, this is great! I found a few nitpicks, otherwise I don't have anything to add

File: src/dfx/src/lib/network/id.rs (hunk 1 above)
Comment (ericswanson-dfinity) [reply src/dfx/src/lib/network/id.rs:12]: Thanks!

File: docs/cli-reference/dfx-start.md [hunk 2]
```diff
@@ -57,3 +57,32 @@ You can view the current process identifier (`pid`) for the local canister execu
...
 more .dfx/pid
 ```
+
+## Local Server Configuration
+
+### The Shared Local Network
+
+By default, `dfx start` manages a single replica that is independent of any given local project.  Running `dfx deploy` and other commands will manage canisters on this single local network, in the same way that `dfx deploy --network ic` deploys separate projects to mainnet.
+
+If run from outside any dfx project, or if dfx.json does not define the `local` network, then `dfx start` looks for the `local` network definition in `$HOME/.config/dfx/networks.json`.  If this file does not exist or does not contain a definition for the `local` network, then dfx uses the following default definition:
```
Comment (sesi200) [inline docs/cli-reference/dfx-start.md:67]:
```suggestion
If run from outside any dfx project, or if dfx.json does not define the `local` network, then `dfx start` looks for the `local` network definition in `$HOME/.config/dfx/networks.json`. If this file does not exist or does not contain a definition for the `local` network, then dfx uses the following default definition:
```

File: docs/cli-reference/dfx-start.md [hunk 3]
```diff
@@ -57,3 +57,32 @@ You can view the current process identifier (`pid`) for the local canister execu
...
+
+dfx stores data for the shared local network in one of the following locations, depending on your operating system:
+- `$HOME/.local/share/dfx/network/local` (Linux)
+- `$HOME/Library/Application Support/org.dfinity.dfx/network/local` (Macos)
+
+### Project-Specific Local Networks
+
+If dfx.json defines the `local` network, then `dfx start` will use this definition and store network data files under `\<project dir\>/.dfx/network/local`. 
+
+Such project-specific networks are deprecated, slated to be removed after February 2023.  We encourage you to remove any definitions of the `local` network from your project's dfx.json file and instead use the default shared local network.
```
Comment (sesi200) [inline docs/cli-reference/dfx-start.md:86]:
```suggestion
Such project-specific networks are deprecated, slanted to be removed after February 2023.  We encourage you to remove any definitions of the `local` network from your project's dfx.json file and instead use the default shared local network.
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,37 @@
...
+See [Local Server Configuration](docs/cli-reference/dfx-start.md#local-server-configuration) for details.
+
+dfx now stores data and control files in one of three places, rather than directly under `.dfx/`:
+- `.dfx/network/local` (for projects in which dfx.json defines the local network)
+- `$HOME/.local/share/dfx/network/local` (for the shared local network on Linux)
+- `$HOME/Library/Application Support/org.dfinity.dfx/network/local` (for the shared local network on MacOS)
+
+There is also a new configuration file: `$HOME/.config/dfx/networks.json`.  Its [schema](docs/networks-json-schema.json) is the same as the `networks` element in dfx.json.  Any networks you define here will be available from any project, unless a project's dfx.json defines a network with the same name.  See [The Shared Local Network](docs/cli-reference/dfx-start.md#the-shared-local-network) for the default definitions that dfx provides if this file does not exist or does not define a `local` network.
+
+### feat: dfx schema --for networks/json
```
Comment (sesi200) [inline CHANGELOG.md:30]:
```suggestion
### feat: dfx schema --for networks
```

**Review by sesi200 — APPROVED** [review]
I think this will be the best-documented feature we've released in quite a while!

File: docs/cli-reference/dfx-start.md (hunk 3 above)
Comment (ericswanson-dfinity) [reply docs/cli-reference/dfx-start.md:86]: "slanted" wouldn't make sense here.  "slated" means planned.  I'll change this to "planned."

https://dictionary.cambridge.org/us/dictionary/english/slated

> be slated
 
> to be [expected](https://dictionary.cambridge.org/us/dictionary/english/expected) to [happen](https://dictionary.cambridge.org/us/dictionary/english/happen) in the [future](https://dictionary.cambridge.org/us/dictionary/english/future) or to be [expected](https://dictionary.cambridge.org/us/dictionary/english/expected) to be or do something in the [future](https://dictionary.cambridge.org/us/dictionary/english/future)

**Pull request merged**